presto-common/src/main/java/com/facebook/presto/common/block/DictionaryBlock.java

```diff
@@ -36,6 +36,21 @@
 import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
 
+/**
+ * A dictionary holds positionCount values of arbitrary types. Uusally some of these values are repeated,
+ * and the wraps an undelrying delegate block without any (or at least fewer) repeated values.
+ * This delegate block is called the "dictionary".
+ * The ids array contains positionCount indexes into the underlying delegate block.
+ * When value N is requested from this block, instead of returning the value directly
+ * it looks up the index of value N at ids[N]; then it returns the value in dictionary[ids[N]].
+ * This compresses data when the same value repeats at multiple locations.
+ *
+ * It is not necessarily the case that every id in the ids array is a valid position in the block.
+ * You can specify an offset in the ids array to indicate that IDs are only stored from that position forward.
+ * Furthermore, if the ids array is longer than offset+positionCount, then extra values to the right are not valid.
+ * That is, IDs are stored in a range of the array from offset to offset+positionCount-1 (inclusive).
```

Comment:
```suggestion
 * That is, ids are stored in a range of the array from offset to offset+positionCount-1 (inclusive).
```

Reply:
I think this last sentence is all we need for this paragraph.  i.e., let's just document that there is an offset as well, and the combination of the offset plus the value in the ids array retrieves the value from the block.

Perhaps even consider removing it altogether and documenting this behavior in the base interface (`#getOffsetBase`).  It's used to define regions over existing blocks, which is shared among all block types, not just dictionaries.

Reply:
I personally found this setup very confusing and it took me awhile to unpack it, so I think extra explanation is warranted. Beware the curse of knowledge.

Reply:
How about instead just documenting it at the base interface?  This method is not special to dictionary blocks, but the documentation only existing here might give one the impression that it is.

Reply:
I'm a little confused. This isn't method documentation. It's class documentation. And it's specifically about the ids array which is unique to DictionaryBlock.

Reply:
Ping. I'd like to finish this one off.